feat(EFB): Simbrief Loadsheet/OFP Navigation

ItzARz opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->

- [x] **Feature:** Navigation through the different Loadsheet/OFP sections in the EFB, improving the reading speed and efficiency.
- [x] It supports these OFP Layouts: [LIDO, AAL, ACA, BAW, DAL, DLH, EIN, EZY, JBU, QFA, THY, BER, KLM]

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make your best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->

https://github.com/user-attachments/assets/d407a8ab-7cfe-47a7-bfdc-5b0d2cd1a61f

![Loadsheet Navigation with EZY OFP Layout](https://github.com/user-attachments/assets/e8cee49f-69b4-43d1-991e-7dd3a851d408)

## References
<!-- You should be making changes based on some kind of a reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

e.g LIDO Format:
![Screenshot 2024-11-11 181233](https://github.com/user-attachments/assets/f5e53dc3-cf90-4eef-bed3-6b111b6f0837)

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context

OFP Layouts: [UAL-2018, UAL-2012, AFR-2012, AFR-2017] were tested but table formatting gets broken, so i decided to not support them for now. I think the issue is caused by the font scaling feature in the loadsheet.

Discord username (if different from GitHub): itzarz

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->

1. Go to Simbrief and generate a flight (make sure to use the Official FBW A20N/A380 Profile).
2. Import the simbrief data through the EFB.
3. Select the **Dispatch** option in the EFB sidebar.
4. Select a desired section and ensure if the feature is working.
5. Compare with Simbrief PDF Sections and its content. 
6. Repeat all the steps with every supported OFP Layout.

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo**, **flybywire-aircraft-a380-842 (4K)** or **flybywire-aircraft-a380-842 (8K)** download link at the bottom of the page

ItzARz committed:
feat(efb): simbrief loadsheet navigation

ItzARz committed:
fix: disable pointer events when loadsheet nav menu is hidden

ItzARz committed:
fix: renamed states setter to match convention

ItzARz committed:
refactor: replaced dropdown with existing SelectInput component.

ItzARz committed:
added error message when loadsheet section not found.

ItzARz committed:
Update CHANGELOG.md

ItzARz committed:
refactor: support for different loadsheet layouts

ItzARz committed:
AFR 2017 Layout support deprecated.

ItzARz committed:
factor: improved load sheet section filtering (Maps missing fixed) and clean up HTML styles.

ItzARz committed:
removed useCallback from filterLoadsheetSection

ItzARz committed:
reset selected section after importing new simbrief data.

ItzARz committed:
added missing Maps section

ItzARz committed:
added sections for KLM and BER OFP Layouts.

ItzARz commented:
I'm not sure sbout creating an interface for the [ofpLayouts.json](https://github.com/flybywiresim/aircraft/blob/716a51e59f7e0cc430664635d6b4a94f826fb25a/fbw-common/src/systems/instruments/src/EFB/Apis/Simbrief/ofpLayouts.json) content and moving the [filterLoadsheetSection](https://github.com/flybywiresim/aircraft/blob/716a51e59f7e0cc430664635d6b4a94f826fb25a/fbw-common/src/systems/instruments/src/EFB/Dispatch/Pages/LoadsheetPage.tsx#L90) function elsewhere, rather than keeping it in the LoadSheetWidget component. However, I would like to hear the opinion of someone with more experience.

I'm open to any suggestions.

2hwk commented on fbw-common/src/systems/instruments/src/EFB/Dispatch/Pages/LoadsheetPage.tsx:
```diff
@@ -75,12 +87,72 @@ export const LoadSheetWidget = () => {
+        nextSection = currentIndex + 2 < loadsheetSections.length ? loadsheetSections[currentIndex + 2] : null;
+      }
+
+      const sectionMarkerPattern = new RegExp(
+        `<!--BKMK///${section}///\\d-->[\\s\\S]*?(?=<!--BKMK///${nextSection || ''}///\\d-->|$)`,
+        'g',
+      );
```
Add some documentation to explain what this regex does in a comment

ItzARz replied:
The loadsheet received from Simbrief is in HTML format, for example:

You can see a couple of the section-markers (in green) in the OFP.
![image](https://github.com/user-attachments/assets/3d4d61a0-9432-47a9-9f2b-ed4c6b3538bf)

Using `[\\s\\S]*?` the regex captures all content, including line breaks, between two section markers
`<!--BKMK///${section}///\\d-->` (the selected section by the user) 
and 
`<!--BKMK///${nextSection || ''}///\\d-->` (next section, if it exists)

tracernz replied:
You can actually just use the DOM API rather than trying to parse it as a string. Comment nodes are nodeType 8 https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType

ItzARz commented:
I accidentally did a force-push, my bad 🤕

2hwk committed:
Merge branch 'master' into feat/EFB-Loadsheet-Navigation

2hwk commented on fbw-common/src/systems/instruments/src/EFB/Apis/Simbrief/ofpLayouts.json:
```diff
@@ -0,0 +1,174 @@
+    "THY": [
+        "All",
+        "OFP",
+        "ATC Flight Plan",
+        "Additional Info",
+        "Runway Analysis",
+        "Airport WX List",
+        "NOTAM",
+        "Company NOTAM",
+        "Maps"
+    ],
+    "KLM": [
+        "All",
+        "Dispatch Briefing",
+        "Administrator Notes",
+        "OFP Flight Particulars",
+        "OFP Navigation Part",
+        "OFP Alternate Summary",
+        "Air Traffic Control",
+        "Lido Briefing",
+        "Runway Analysis",
+        "UAD Charts",
+        "Airports Weather",
+        "NOTAM",
+        "Crew Alert"
+    ],
+    "BER": [
+        "All",
+        "Summary",
+        "Flight Log",
+        "Wind Information",
+        "ATC Flight Plan",
+        "Additional Info",
+        "Runway Analysis",
+        "Weather",
+        "Notam",
+        "Crew Alert & Bulletin",
+        "Maps"
+    ]
+}
```
Long term I'd like to move this into either base/config or as part of aircraft context (I've not decided yet which would fit better for this). Having loose .json around the repository is hard for maintainers to keep track of.

ItzARz committed:
Merge branch 'master' into feat/EFB-Loadsheet-Navigation

heclak commented:
The PR seems to be missing the builds for QA.

ItzARz commented:
> The PR seems to be missing the builds for QA.

Builds are in progress, thanks for letting know. It will be ready in 30mins i think.

LeDerpLegend commented:
<!-- !! DO NOT REMOVE THIS LINE !! - QUALITY ASSURANCE REPORT -->
Quality Assurance Trainee Report

Discord Username  : LeDerpLegend
Object of testing : #9527 
Aircraft               : A32NX/A380X
Tier of Testing     : 1
Date                   : 28/11/2024

Testing Process: 
Spawned in A32NX at KLAS
Imported Simbrief profile KLAS/KJFK, it auto loaded
Selected the Dispatch option in the EFB sidebar
Selected a desired section and ensured if the feature is working
Compared with Simbrief PDF Sections and its content
Loaded a new Simbrief flight from KLAX/KJFK and loaded it into the EFB and Repeated previous three steps.

Repeated steps above with A380X except with plans from KLAX/KPHX and KSAN/CYYZ

Testing Results: 
**PASSED**

Negatives:       
If I have something deselected in Simbrief under selections (such as NOTAMS), selecting NOTAMS on the EFB dispatch will display the default ALL section. Would recommend displaying a Info not Available, or Section not Enabled in Simbrief. Not a major issue but would help those who are confused. 

Conclusions: 
Working as intended for section selection. 
EFB warned me about incorrect aircraft profile when I had the A32NX for the A380X. Nice

Maximilian-Reuter commented:
Sorry for coming in so late in the progress here, but I was wondering why the decision was made to limit the scrolling to the selected section ? I would have expected the function to kind of work like the sections of a webpage where all it does is scroll you down to that area. It can be seen in your example video. Going to different sections especially consecutive ones requires additional clicks now that it didnt before, which I personally consider a reduction in UX.

Also I echo the Negatives that lederp talks about. Would it be hard to only show sections that actually exist in the OFP?

LeDerpLegend commented:
> Sorry for coming in so late in the progress here, but I was wondering why the decision was made to limit the scrolling to the selected section ? I would have expected the function to kind of work like the sections of a webpage where all it does is scroll you down to that area. It can be seen in your example video. Going to different sections especially consecutive ones requires additional clicks now that it didnt before, which I personally consider a reduction in UX.
> 
> Also I echo the Negatives that lederp talks about. Would it be hard to only show sections that actually exist in the OFP?

One reason I could see tabs that limit sections is that it can make it easier to digest stuff rather than looking at it and possibly going into extra information. Can help newer pilots or those who want things simple. Plus prevents accidental scrolling off a section. But I could see that version working as well, I personally find this version works for me.

ItzARz commented:
My bad for the late reply i was a bit busy.

Thank you for the feedback Chaoz.

Regarding this:
> why the decision was made to limit the scrolling to the selected section ? 

Lederp already provided a great explanation:
> it can make it easier to digest stuff rather than looking at it and possibly going into extra information. Can help newer pilots or those who want things simple. Plus prevents accidental scrolling off a section. 

When designing this feature, I considered two approaches. One (automatic scrolling) was the method you mentioned:
> I would have expected the function to kind of work like the sections of a webpage where all it does is scroll you down to that area.

However, I found that implementing automatic scrolling was more challenging to me.

Anyway, I’m open to revisiting this and exploring the possibility of using automatic scrolling instead of limiting the section content. If this approach provides a better user experience, I’d be happy to implement it. I would just need further clarification or guideline to do it.

As an alternative, I could add two small arrow buttons in a corner to quickly navigate to the section above or below the current one.
Regarding the negative point in the QA test—good catch, I didn’t account for that case, but I’ll make sure it gets fixed.

ItzARz committed:
option added in EFB to activate/deactivate the loadsheet navigation and improved UX.

Now there is two arrow buttons to quickly navigation between loadsheet sections.

ItzARz committed:
added: message when a section is not available.

ItzARz committed:
code clean up.

ItzARz commented:
Quick video showcasing the navigation with the new buttons to switch "faster" between sections. 
https://github.com/user-attachments/assets/f8e5cb1a-3580-4c87-ac92-9723a39cffa1

Btw, Localazy keys needed.

tracernz commented on .github/CHANGELOG.md:
```diff
@@ -71,6 +71,7 @@
 1. [FMS] Fix Pause at T/D not working in selected speed - @BlueberryKing (BlueberryKing)
 1. [MISC] Replaced brake temperature simulation with physics based model of brakes - @Gurgel100 (Pascal)
 1. [A32NX/MCDU] Suppress TMPY FPLN when no modifications made in airways page - @robertxing2004 (robeet)
+1. [EFB/SIMBRIEF] Added loadsheet navigation - @ItzARz (ARz)
```
Please move to 0.14.0.

ItzARz committed:
Merge branch 'master' into feat/EFB-Loadsheet-Navigation

ItzARz committed:
Update CHANGELOG.md

Jaronics commented:
<!-- !! DO NOT REMOVE THIS LINE !! - QUALITY ASSURANCE REPORT -->
<!-- Refer to the https://docs.flybywiresim.com/dev-corner/qa-process/#reporting-on-tests for details about this template -->

<!-- Please use one or the other depending on your Discord Role -->
Quality Assurance Trainee Report

Discord Username  : Jaronics <!-- Use your username, not your display name! -->
Object of testing : #9527  <!-- PR Number -->
Aircraft          : A32NX <!-- Choose only one -->
Simulator         : 2020 <!-- Choose only one --> 
Tier of Testing   : 1
Date              : 04/07/2025 <!-- Date of test in DD/MM/YYYY format -->

Testing Process: 
Loaded up a flight plan and tested every supported layout. 

LIDO: Crew Alert & Bulletin shows the names of the map charts at the bottom. The last line is cut off. (pic 1) (could be due to zoom level?)

AAL: 1. Wrong formatting. Many of the sections don't exist in the AAL layout, causing an error. (pic 2) Example: NOTAMs is instead 
            called NOTAM Briefing.
        2. On the normal ALL page, the zoom function does not work.
        
ACA: Map names are displayed at the bottom of Runway Analysis page.

BAW: 1. Last line of Weights and Performance is cut off.
         2. Crew Information once again shows the names of the maps at the end.
   
DAL: 1. The first portion of the Flight Log is found in DX Remarks instead. (pic 4 and 5)
        2. Map names are shown at the end of Runway Analysis.
  
DLH: Map names are shown at the end of NOTAM.

EIN: 1. Flight Log and Flight Log to the alternate are split in the EIN OFP. Here, both are grouped together under the normal log.
       2. Map names are shown at the end of Crew Information.

EZY: 1. Beginning of Flight Log is displayed at the end of Summary and Fuel. 
        2. Map names are shown at the end of Company NOTAM.
     
JBU: Map names are shown at the end of Runway Analysis.

QFA: 1. The first line of Weather and NOTAMs is included in Computer Flight Plan instead. 
        2. Map names are displayed at the end of Runway Analysis.
    
THY: Company NOTAM includes the names of the maps.

BER: Crew Alert & Bulletin shows the names of the map charts at the bottom. The last line is cut off. (pic 1)

KLM: 1. OFP Alternate Summary does not include the routing, as it does in the PDF. That is wrongly in the OFP Navigation Part.
         2. Runway Analysis includes the map names at the end.
         4. The last line of Crew Alert is cut off, if fully zoomed in. Probably also the case with the other cut-off sections.

<!-- List the process - Required section --> 

Testing Results: 
Not Passed <!-- Required section --> 

Negatives: 
1. Loading in, I needed to turn the setting OFF and ON to make the feature work.  
2. Scroll thingy goes partly outside the screen when scrolling all the way down in the NOTAM section. (pic 3)    
<!-- Ommit this section if none are found -->

Conclusions: 
AAL needs a rework. All others work for the most part, only with some small positioning errors.

I agree with @Maximilian-Reuter that I would prefer if the whole OFP remained available to scroll instead of just scrolling to a specific part of it and locking the user there.
<!-- Ommit section if nothing to add -->

Media: 
1: 
![grafik](https://github.com/user-attachments/assets/71e98cdb-cf51-426f-acde-8d91dcc5ed2b)
2:
![grafik](https://github.com/user-attachments/assets/6186bfb6-3bfc-4fa2-a1b4-2a3246261a87)
3:
![grafik](https://github.com/user-attachments/assets/025dbbae-c2c5-4831-86c3-ea55335b2c14)
4:
![grafik](https://github.com/user-attachments/assets/2d92b595-5603-46c5-8652-5c11bc53703a)
5:
![grafik](https://github.com/user-attachments/assets/3639e7c1-34de-4fa0-87ca-f9fbc7dc30cb)

<!-- Ommit this section if no media to add -->
